### PR TITLE
fix: user context refactor

### PR DESCRIPTION
**Pull request by sattvikc**
## Summary of change

(A few sentences about this PR)

## Related issues

-   Link to issue1 here
-   Link to issue1 here

## Test Plan

(Write your test plan here. If you changed any code, please provide us with clear instructions on how you verified your changes work. Bonus points for screenshots and videos!)

## Documentation changes

(If relevant, please create a PR in our [docs repo](https://github.com/supertokens/docs), or create a checklist here highlighting the necessary changes)

## Checklist for important updates

-   [ ] Changelog has been updated
-   [ ] `coreDriverInterfaceSupported.json` file has been updated (if needed)
    -   Along with the associated array in `lib/ts/version.ts`
-   [ ] `frontendDriverInterfaceSupported.json` file has been updated (if needed)
-   [ ] Changes to the version if needed
    -   In `package.json`
    -   In `package-lock.json`
    -   In `lib/ts/version.ts`
-   [ ] Had run `npm run build-pretty`
-   [ ] Had installed and ran the pre-commit hook
-   [ ] Issue this PR against the latest non released version branch.
    -   To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    -   If no such branch exists, then create one from the latest released branch.
-   [ ] If have added a new web framework, update the `add-ts-no-check.js` file to include that
-   [ ] If added a new recipe / api interface, then make sure that the implementation of it uses NON arrow functions only (like `someFunc: function () {..}`).
-   [ ] If added a new recipe, then make sure to expose it inside the recipe folder present in the root of this repo. We also need to expose its types.

## Remaining TODOs for this PR

-   [ ] Item1
-   [ ] Item2

### CODE REVIEW

File: lib/ts/recipeModule.ts [hunk 1]
```diff
@@ -56,18 +58,28 @@ export default abstract class RecipeModule {
...
+                    return { id: currAPI.id, tenantId: finalTenantId === undefined ? DEFAULT_TENANT_ID : tenantId };
                 } else if (
                     remainingPath !== undefined &&
                     this.appInfo.apiBasePath
                         .appendPath(currAPI.pathWithoutApiBasePath)
                         .equals(this.appInfo.apiBasePath.appendPath(remainingPath))
                 ) {
-                    return { id: currAPI.id, tenantId };
+                    const finalTenantId = mtRecipe.recipeInterfaceImpl.getTenantId({
+                        tenantIdFromFrontend: DEFAULT_TENANT_ID,
```
Comment (rishabhpoddar) [inline lib/ts/recipeModule.ts:79]:
```suggestion
                        tenantIdFromFrontend: tenantId,
```

File: lib/ts/recipeModule.ts [hunk 2]
```diff
@@ -56,18 +58,28 @@ export default abstract class RecipeModule {
...
+                    });
+                    return { id: currAPI.id, tenantId: finalTenantId === undefined ? DEFAULT_TENANT_ID : tenantId };
                 } else if (
                     remainingPath !== undefined &&
                     this.appInfo.apiBasePath
                         .appendPath(currAPI.pathWithoutApiBasePath)
                         .equals(this.appInfo.apiBasePath.appendPath(remainingPath))
                 ) {
-                    return { id: currAPI.id, tenantId };
+                    const finalTenantId = mtRecipe.recipeInterfaceImpl.getTenantId({
```
Comment (rishabhpoddar) [inline lib/ts/recipeModule.ts:78]: make the input and output type of getTenantId non optional

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: lib/ts/recipe/dashboard/api/validateKey.ts [hunk 3]
```diff
@@ -13,12 +13,11 @@
  * under the License.
  */
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply lib/ts/recipe/dashboard/api/validateKey.ts:15]: Done

File: lib/ts/framework/awsLambda/framework.ts [hunk 4]
```diff
@@ -22,7 +22,7 @@ import type {
     Callback,
 } from "aws-lambda";
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply lib/ts/framework/awsLambda/framework.ts:23]: Done

File: lib/ts/recipeModule.ts (hunk 2 above)
Comment (sattvikc) [reply lib/ts/recipeModule.ts]: Done

**Pull request merged**